cherab/generomak/machine/plasma.py

```diff
@@ -0,0 +1,213 @@
+    mesh_interp["electron"]["density"] = ne
+
+    for elem_name, elem_data in profiles["composition"].items():
+        for stage, stage_data in elem_data.items():
+
+            t = Discrete2DMesh.instance(te, stage_data["temperature"], limit=False)
+            n = Discrete2DMesh.instance(te, stage_data["density"], limit=False)
+
+            mesh_interp["composition"][elem_name][stage]["temperature"] = t
+            mesh_interp["composition"][elem_name][stage]["density"] = n
+            mesh_interp["composition"][elem_name][stage]["element"] = stage_data["element"]
+
+    
+    return mesh_interp.freeze()
+    
+def get_edge_distributions():
+    """
+    Provides Generomak edge Maxwellian distribution of plasma species
+
+    :return: Dictionary holding instances of Maxwellian distributions for plasma species.
+    """
+
+    mesh_interp = get_edge_interpolators()
+
+    zero_vector = Vector3D(0, 0, 0)
+
+    dists = RecursiveDict()
+
+    n3d = AxisymmetricMapper(mesh_interp["electron"]["density"])
+    t3d = AxisymmetricMapper(mesh_interp["electron"]["temperature"])
+
+    dists["electron"] = Maxwellian(n3d, t3d, zero_vector, electron_mass)
+
+    for elem_name, elem_data in mesh_interp["composition"].items():
+        for stage, stage_data in elem_data.items():
+            
+            n3d = AxisymmetricMapper(stage_data["density"])
+            t3d = AxisymmetricMapper(stage_data["temperature"])
+            mass = stage_data["element"].atomic_weight * atomic_mass
+            dists["composition"][elem_name][stage]["distribution"] = Maxwellian(n3d, t3d, zero_vector, mass)
+            dists["composition"][elem_name][stage]["element"] = stage_data["element"]
+
+    return dists.freeze()
+
+def get_edge_plasma(atomic_data=None):
+    """
+    Provides Generomak Edge plasma.
+
+    :param atomic_data: Instance of AtomicData, default isOpenADAS()
+    :return: populated Plasma object
+    """
+    
+    #create or check atomic_data
+    if atomic_data is not None:
+     if not isinstance(atomic_data, AtomicData):
+         raise ValueError("atomic_data has to be of type AtomicData")   
+    else:
+        atomic_data = OpenADAS()
+
+    # base plasma geometry on mesh vertices
+    profiles_dir = os.path.join(os.path.dirname(__file__), "data/plasma/edge")
+    path = os.path.join(profiles_dir, "mesh.json")
+    with open(path, "r") as fhl:
+        mesh = json.load(fhl)
+
+    vertex_coords = np.asarray(mesh["vertex_coords"])
+    r_range = (vertex_coords[:, 0].min(), vertex_coords[:, 0].max())
+    z_range = (vertex_coords[:, 1].min(), vertex_coords[:, 1].max())
+    plasma_height = z_range[1] - z_range[0]
+
+    padding = 1e-3 #enlarge plasma geometry for safety
+
+    outer_column = Cylinder(radius=r_range[1] + padding, height=plasma_height + 2 * padding)
+    inner_column = Cylinder(radius=r_range[0] - padding, height=plasma_height + 4 * padding)
+    inner_column.transform = translate(0, 0, -2 * padding)
```

Comment:
Can you please clarify how adding this padding improves the safety? In `RayTransferBox` and `RayTransferCylinder`, the geometry of the bounding primitive needed to be reduced a little to avoid out-of-range errors for cell indices. But what could go wrong here?

Reply:
The interpolators are set to give 0 outside the triangles, which should make it save to extend the bounding primitive. I wanted to make sure that the bounds include the whole mesh, some SOLPS triangles in the divertor region can be very small, but you are right that not using padding would not influence the results, especially the demos. Should I remove it?

Reply:
The bounds `r_range` and `z_range` already include the whole mesh no matter how small the triangles are. I would remove this code because it can be confusing, especially in adding different number of padding to the different parameters of the primitives. But maybe I'm missing something, so let's ask for @jacklovell's opinion.

Reply:
If the padding is unnecessary I would remove it, and keep the function as simple as possible.

Reply:
Removed padding to enlarge the plasma primitive. Padding stayed in the case of inner column to avoid subtraction of parallel plains